// react-native-inner-shadow/lib/commonjs/components/CornerRadii.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CornerRadii = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNativeSkia = require("@shopify/react-native-skia");
var _utils = require("../utils.js");
var _constants = require("../constants.js");
var _jsxRuntime = require("react/jsx-runtime");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const CornerRadii = exports.CornerRadii = /*#__PURE__*/_react.default.memo(function CornerRadii({
  width,
  height,
  style,
  children,
  backgroundColor
}) {
  const rrct = _react.default.useMemo(() => {
    const {
      topLeftRadius,
      topRightRadius,
      bottomRightRadius,
      bottomLeftRadius
    } = (0, _utils.getBorderRadius)({
      borderRadius: style?.borderRadius,
      borderTopStartRadius: style?.borderTopStartRadius,
      borderTopLeftRadius: style?.borderTopLeftRadius,
      borderTopEndRadius: style?.borderTopEndRadius,
      borderTopRightRadius: style?.borderTopRightRadius,
      borderBottomStartRadius: style?.borderBottomStartRadius,
      borderBottomLeftRadius: style?.borderBottomLeftRadius,
      borderBottomEndRadius: style?.borderBottomEndRadius,
      borderBottomRightRadius: style?.borderBottomRightRadius
    });
    return {
      rect: {
        x: _constants.CANVAS_PADDING,
        y: _constants.CANVAS_PADDING,
        width: width,
        height: height
      },
      topLeft: {
        x: topLeftRadius,
        y: topLeftRadius
      },
      topRight: {
        x: topRightRadius,
        y: topRightRadius
      },
      bottomRight: {
        x: bottomRightRadius,
        y: bottomRightRadius
      },
      bottomLeft: {
        x: bottomLeftRadius,
        y: bottomLeftRadius
      }
    };
  }, [width, height, style?.borderRadius, style?.borderTopStartRadius, style?.borderTopLeftRadius, style?.borderTopEndRadius, style?.borderTopRightRadius, style?.borderBottomStartRadius, style?.borderBottomLeftRadius, style?.borderBottomEndRadius, style?.borderBottomRightRadius]);
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNativeSkia.RoundedRect, {
    rect: rrct,
    color: backgroundColor,
    children: children
  });
});
