// react-native-inner-shadow/lib/module/constants.js
"use strict";

import { StyleSheet } from 'react-native';
const CANVAS_PADDING = 50;
const BACKGROUND_COLOR = '#FFFFFF';

// These two scales are opposite each other to create a "reflected light" effect.
const SHADOW_OFFSET_SCALE = 2.5;
const REFLECTED_LIGHT_OFFSET_SCALE = 2;
const INITIAL_DEPTH = 2;
const SHADOW_OPACITY = 0.3;
const SHADOW_RADIUS = 3;
const SHADOW_BLUR = 2;
const SHADOW_ELEVATION = 3;
const REFLECTED_LIGHT_BLUR = 3;
const SHADOW_COLOR = '#2F2F2FBC';
const REFLECTED_LIGHT_COLOR = '#FFFFFF4D';
const DAMPING_DURATION = 150;
const DAMPING_RATIO = 0.8;
const IS_REFLECTED_LIGHT_ENABLED = true;
const COMMON_STYLES = StyleSheet.create({
  canvasContainer: {
    backgroundColor: 'transparent'
  },
  canvasWrapper: {
    backgroundColor: 'transparent'
  },
  canvas: {
    position: 'absolute',
    left: -CANVAS_PADDING,
    top: -CANVAS_PADDING,
    backgroundColor: 'transparent'
  }
});
export { CANVAS_PADDING, BACKGROUND_COLOR, INITIAL_DEPTH, SHADOW_OPACITY, SHADOW_RADIUS, SHADOW_BLUR, REFLECTED_LIGHT_BLUR, SHADOW_COLOR, REFLECTED_LIGHT_COLOR, DAMPING_DURATION, DAMPING_RATIO, IS_REFLECTED_LIGHT_ENABLED, SHADOW_OFFSET_SCALE, REFLECTED_LIGHT_OFFSET_SCALE, SHADOW_ELEVATION, COMMON_STYLES };
