// react-native-inner-shadow/lib/commonjs/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SHADOW_RADIUS = exports.SHADOW_OPACITY = exports.SHADOW_OFFSET_SCALE = exports.SHADOW_ELEVATION = exports.SHADOW_COLOR = exports.SHADOW_BLUR = exports.REFLECTED_LIGHT_OFFSET_SCALE = exports.REFLECTED_LIGHT_COLOR = exports.REFLECTED_LIGHT_BLUR = exports.IS_REFLECTED_LIGHT_ENABLED = exports.INITIAL_DEPTH = exports.DAMPING_RATIO = exports.DAMPING_DURATION = exports.COMMON_STYLES = exports.CANVAS_PADDING = exports.BACKGROUND_COLOR = void 0;
var _reactNative = require("react-native");
const CANVAS_PADDING = exports.CANVAS_PADDING = 50;
const BACKGROUND_COLOR = exports.BACKGROUND_COLOR = '#FFFFFF';

// These two scales are opposite each other to create a "reflected light" effect.
const SHADOW_OFFSET_SCALE = exports.SHADOW_OFFSET_SCALE = 2.5;
const REFLECTED_LIGHT_OFFSET_SCALE = exports.REFLECTED_LIGHT_OFFSET_SCALE = 2;
const INITIAL_DEPTH = exports.INITIAL_DEPTH = 2;
const SHADOW_OPACITY = exports.SHADOW_OPACITY = 0.3;
const SHADOW_RADIUS = exports.SHADOW_RADIUS = 3;
const SHADOW_BLUR = exports.SHADOW_BLUR = 2;
const SHADOW_ELEVATION = exports.SHADOW_ELEVATION = 3;
const REFLECTED_LIGHT_BLUR = exports.REFLECTED_LIGHT_BLUR = 3;
const SHADOW_COLOR = exports.SHADOW_COLOR = '#2F2F2FBC';
const REFLECTED_LIGHT_COLOR = exports.REFLECTED_LIGHT_COLOR = '#FFFFFF4D';
const DAMPING_DURATION = exports.DAMPING_DURATION = 150;
const DAMPING_RATIO = exports.DAMPING_RATIO = 0.8;
const IS_REFLECTED_LIGHT_ENABLED = exports.IS_REFLECTED_LIGHT_ENABLED = true;
const COMMON_STYLES = exports.COMMON_STYLES = _reactNative.StyleSheet.create({
  canvasContainer: {
    backgroundColor: 'transparent'
  },
  canvasWrapper: {
    backgroundColor: 'transparent'
  },
  canvas: {
    position: 'absolute',
    left: -CANVAS_PADDING,
    top: -CANVAS_PADDING,
    backgroundColor: 'transparent'
  }
});
