// react-native-inner-shadow/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "LinearShadowPressable", {
  enumerable: true,
  get: function () {
    return _ShadowPressable.LinearShadowPressable;
  }
});
Object.defineProperty(exports, "LinearShadowToggle", {
  enumerable: true,
  get: function () {
    return _ShadowToggle.LinearShadowToggle;
  }
});
Object.defineProperty(exports, "LinearShadowView", {
  enumerable: true,
  get: function () {
    return _ShadowView.LinearShadowView;
  }
});
Object.defineProperty(exports, "ShadowPressable", {
  enumerable: true,
  get: function () {
    return _ShadowPressable.ShadowPressable;
  }
});
Object.defineProperty(exports, "ShadowToggle", {
  enumerable: true,
  get: function () {
    return _ShadowToggle.ShadowToggle;
  }
});
Object.defineProperty(exports, "ShadowView", {
  enumerable: true,
  get: function () {
    return _ShadowView.ShadowView;
  }
});
Object.defineProperty(exports, "computeShadowProperties", {
  enumerable: true,
  get: function () {
    return _utils.computeShadowProperties;
  }
});
Object.defineProperty(exports, "getBackgroundColor", {
  enumerable: true,
  get: function () {
    return _utils.getBackgroundColor;
  }
});
Object.defineProperty(exports, "getLinearDirection", {
  enumerable: true,
  get: function () {
    return _utils.getLinearDirection;
  }
});
Object.defineProperty(exports, "isLinearProps", {
  enumerable: true,
  get: function () {
    return _utils.isLinearProps;
  }
});
Object.defineProperty(exports, "useAnimatedOffset", {
  enumerable: true,
  get: function () {
    return _useAnimatedOffset.useAnimatedOffset;
  }
});
Object.defineProperty(exports, "useShadowProperties", {
  enumerable: true,
  get: function () {
    return _useShadowProperties.useShadowProperties;
  }
});
var _utils = require("./utils.js");
var _useShadowProperties = require("./hooks/useShadowProperties.js");
var _useAnimatedOffset = require("./hooks/useAnimatedOffset.js");
var _ShadowView = require("./components/ShadowView.js");
var _ShadowPressable = require("./components/ShadowPressable.js");
var _ShadowToggle = require("./components/ShadowToggle.js");
